### PR TITLE
Reviewing HW submission from dopefish2112

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

File: hw2/dopefish2112.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (westonpace) [inline hw2/dopefish2112.cs:5]: I think VS might add a bunch of `using` statements by default but some of these (e.g. this one) are not needed.

File: hw2/dopefish2112.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+using System.Diagnostics;
+
+namespace CSharp_Zero_to_Hero
+{
+    class Program
+    {
+        /*
```
Comment (westonpace) [inline hw2/dopefish2112.cs:12]: Professionalism nit: Try and clean up comments like this before submitting things.

File: hw2/dopefish2112.cs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        static string c2M = "please type a word the append to the end of the array. . .";
+        */
+
+        static void Main(string[] args)
+        {
+            
+            string[] myArray = { "herp", "derp", "twerp", "gerp", "zerp", "furp" };
+            string uI1;
+            int uI2;
+            string c1M = "Please type a word the append to the begining of the array. . .";
```
Comment (westonpace) [inline hw2/dopefish2112.cs:26]: Nit: If a string literal is only used in exactly one place in your code then separating them into dedicated variables harms readability IMO.

More significant: If you are going to proceed this way then you need to fix the variable names.  `c3M` is meaningless.  Use something like `RANDOM_ADD_PROMPT` (convention for constants is to use ALL_CAPS but the more meaningful name is more important than the all caps)

File: hw2/dopefish2112.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            string c3M = "Please type a word to append randomly within the array";
+            string c4M = "Deleting the 1st element in the Array. . .";
+            string c5M = "Deleting the last element in the Array. . .";
+            string c6M = "Please choose an index number to delete. . .";
+
+            logIn();
+            print(myArray);
+            waitForUser();
+
+            sortArray(ref myArray);
```
Comment (westonpace) [inline hw2/dopefish2112.cs:37]: What's with all of the `ref`?  Can you think of a way to make this work without using the `ref` keyword?

File: hw2/dopefish2112.cs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                    Console.WriteLine("Login failure. Press enter to exit.");
+                    Console.ReadLine();
+                    Environment.Exit(0);    
+            }
+ 
+            else
+            {
+                Console.WriteLine("Login successful");
+            }
+            Console.ReadKey();
```
Comment (westonpace) [inline hw2/dopefish2112.cs:225]: There was no prompt here and it wasn't clear to me as a user that I needed to press enter.  Maybe add "Please press enter to continue"

File: hw2/dopefish2112.cs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+            Console.WriteLine(c6M);
+            uI2 = Int32.Parse(Console.ReadLine());
+            uI2--;
+            delSelectedElement(ref myArray, uI2);
+            waitForUser();
+            
+        }
+
+        public static void print(string[] i)
```
Comment (westonpace) [inline hw2/dopefish2112.cs:71]: The conventions in C# is to start method names with a capital letter (`Print`, `SortArray`, etc.)

File: hw2/dopefish2112.cs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            delFirstElement(ref myArray);
+            waitForUser();
+
+            Console.WriteLine(c5M);
+            delLastElement(ref myArray);
+            waitForUser();
+
+            Console.WriteLine(c6M);
+            uI2 = Int32.Parse(Console.ReadLine());
+            uI2--;
```
Comment (westonpace) [inline hw2/dopefish2112.cs:65]: Your prompt says `Please choose an index number to delete`.  Either don't use the word `index` (i.e. assume your audience is non-technical) or do not decrement by 1 (i.e. if you are assuming a technical audience then respect the fact that "indices" are 0-based).

Even better, when you print the array do something like...

```
1) herp
2) derp
...
```

Then it is very clear how the user's input matches the data.

File: hw2/dopefish2112.cs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            delSelectedElement(ref myArray, uI2);
+            waitForUser();
+            
+        }
+
+        public static void print(string[] i)
+        {
+            Console.WriteLine("THIS IS THE ORIGINAL ARRAY. . .");
+            foreach (var item in i)
```
Comment (westonpace) [inline hw2/dopefish2112.cs:74]: You print the array like this many times, make a helper method.

File: hw2/dopefish2112.cs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        public static void changeFirstElement(ref string[] i, string j)
+        {
+            Array.Resize(ref i, i.Length + 1);
+            for (int x = i.Length - 1; x > 0; x--)
+            {
+                i[x] = i[x - 1];
+            }
+            i[i.Length - i.Length] = j;
```
Comment (westonpace) [inline hw2/dopefish2112.cs:98]: Can you think of a way to simplify `i.Length - i.Length`?

File: hw2/dopefish2112.cs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+            Array.Sort(i);
+            Console.WriteLine("THIS IS THE SORTED ARRAY. . .");
+            foreach (var item in i)
+            {
+                Console.WriteLine(item);
+            }
+
+        }
+
+        public static void changeFirstElement(ref string[] i, string j)
```
Comment (westonpace) [inline hw2/dopefish2112.cs:91]: Nit: You aren't actually changing any elements.  You are adding a new element.

File: hw2/dopefish2112.cs [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+            foreach (var word in i)
+            {
+                Console.WriteLine(word);
+            }
+        }
+
+        public static void changeRandomElement(ref string[] i, string j)
+        {
+            var rand = new Random();
+            int num = rand.Next(0, i.Length - 1);
```
Comment (westonpace) [inline hw2/dopefish2112.cs:119]: Nit: You can use `rand.Next(0, i.Length)`  The second argument is exclusive.  In math terms we'd say `[0, i.Length)`

File: hw2/dopefish2112.cs [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+                }
+                
+                else if (num == x)
+                {
+                    i[x] = j;
+                }
+
+                else if (num > x)
+                {
+                    break;
```
Comment (westonpace) [inline hw2/dopefish2112.cs:135]: Clever working through the array backwards like this which allows you to use `break`

File: hw2/dopefish2112.cs [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        public static void waitForUser()
+        {
+            Console.WriteLine("Press enter to continue. . .");
+            Console.ReadLine();
+        }
+
+        public static void logIn()
+        {
+            //Login Attempts counter
+            int loginAttempts = 0;
```
Comment (westonpace) [inline hw2/dopefish2112.cs:197]: If you're going to do 3 login attempts you should add a "Invalid username or password" print somewhere.

**Review by westonpace — COMMENTED** [review]
It seems like you preferred `Array.Resize` over creating a new array.  Any particular reasoning?

File: hw2/dopefish2112.cs (hunk 1 above)
Comment (dopefish2112) [reply hw2/dopefish2112.cs:5]: you are correct. i did not refactor this and should have.

File: hw2/dopefish2112.cs (hunk 4 above)
Comment (dopefish2112) [reply hw2/dopefish2112.cs:37]: i could not get it to work without the ref. i wanted to ask about this when we someone had some time. if there is a way to do with with return; possibly?

File: hw2/dopefish2112.cs (hunk 9 above)
Comment (dopefish2112) [reply hw2/dopefish2112.cs:98]: just use hardcode i[0]?

File: hw2/dopefish2112.cs (hunk 11 above)
Comment (dopefish2112) [reply hw2/dopefish2112.cs:119]: so that wouldn't create an index outside of the bounds of the array?

File: hw2/dopefish2112.cs (hunk 4 above)
Comment (AlmightyLks) [reply hw2/dopefish2112.cs:37]: I don't quite understand why you're `ref`'ing an array. Arrays are reference types by nature

File: hw2/dopefish2112.cs (hunk 9 above)
Comment (AlmightyLks) [reply hw2/dopefish2112.cs:98]: Yes

File: hw2/dopefish2112.cs (hunk 11 above)
Comment (AlmightyLks) [reply hw2/dopefish2112.cs:119]: Yes. Also, my two cents:

Create a Random object for the whole class, don't re-create it here everytime.
TLDR; Random is not as random as you might think in .NET 😄

Comment (dopefish2112) [reply hw2/dopefish2112.cs:119]: oh wait. 8th grade math coming back to me.

so this is saying 0 <= x < i.Length ?

File: hw2/dopefish2112.cs (hunk 3 above)
Comment (dopefish2112) [reply hw2/dopefish2112.cs:26]: > It seems like you preferred `Array.Resize` over creating a new array. Any particular reasoning?

It seemed to end up with less code when I did it this way. TBH i just searched the documentation for Array helpers and ran with it. So i guess no. no particular reason.

File: hw2/dopefish2112.cs (hunk 4 above)
Comment (dopefish2112) [reply hw2/dopefish2112.cs:37]: so i was having problems getting the array variable to update when the method ran and i was getting errors without the ref. this is what VS suggested and it made my red squiggles go away.

Comment (westonpace) [reply hw2/dopefish2112.cs:37]: The source of the issue is `Array.Resize`.  If you look at the definition it says `public static void Resize<T> (ref T[]? array, int newSize);`  This covers a number of things Kai hasn't talked about yet so here is the short version.

1) Most functions take stuff in and output based on the inputs.  This would be called a "pure" function (assuming it's static but let's not worry about that for now).  This has been the case for all functions so far.
2) Some functions take stuff in and modify the stuff you passed in instead of returning something.
3) Normally there are rules that work really well for this and we will cover them in OOP.
4) Sometimes you have to break the rules, this is what the `ref` keyword is for.  The method `Array.Resize` uses `ref` to say "I am going to modify the array you are giving me instead of returning a new array".
5) All else sort of follows from that.

The way to work around it would be to not use `Array.Resize` or you could go halfway and do something like...

```
        public static string[] changeLastElement(string[] i, string j)
        {
            var localArr = i;
            Array.Resize(ref localArr, i.Length + 1);
            i[i.Length - 1] = j;
            Console.WriteLine("Here is the new array. . .");
            foreach (var word in i)
            {
                Console.WriteLine(word);
            }
            return localArr;
        }
```

Here we are making a new local variable `localArr` and passing that by reference (`Array.Resize(ref localArr, i.Length+1);`).  However, instead of also taking in a ref array we are taking in a plain array and returning a new array instead.  You would have to change your calls to...

```
             myArray = changeLastElement(myArray, uI1);
```

Comment (westonpace) [reply hw2/dopefish2112.cs:37]: This will be talked about in more detail when we talk about object oriented programming.  If you want to read ahead you can look for information about "reference types vs value types" but I'd recommend waiting.

Comment (westonpace) [reply hw2/dopefish2112.cs:37]: @AlmightyLks You probably guessed it by now but you are right, arrays are reference types.  However, you can use `ref` on a reference type if you want to modify what the reference points to and not just the state of the referenced object.

File: hw2/dopefish2112.cs (hunk 11 above)
Comment (westonpace) [reply hw2/dopefish2112.cs:119]: Yes

File: hw2/dopefish2112.cs (hunk 3 above)
Comment (westonpace) [reply hw2/dopefish2112.cs:26]: Ok, there is sometimes a misconception that Array.Resize is faster than making a new array and copying the elements but it is actually still creating a new array and copying elements under the hood.
